CKAN 2.9 CLI documentation (#5176)
ckan/ckan: doc/maintaining/cli.rst

.. _cli:

======================
Command Line Interface (CLI)
======================

> NB: From CKAN 2.9 onwards the CKAN configuration file is named 'ckan.ini'. Previous names: 'production.ini' and 'development.ini' (plus others) may also still appear in documentation and the software. These legacy names will eventually be phased out. Also from CKAN 2.9 onwards, the `paster`` command used for common CKAN administration tasks has been replaced with the  ``ckan`` command.

Reviewer: It would be more consistent to use `.. note::` Also:
```suggestion
NB: From CKAN 2.9 onwards the CKAN configuration file is named 'ckan.ini'. Previous names: 'production.ini' and 'development.ini' (plus others) may also still appear in documentation and the software. These legacy names will eventually be phased out. Also from CKAN 2.9 onwards, the ``paster`` command used for common CKAN administration tasks has been replaced with the  ``ckan`` command.
```

dataset           Manage datasets.
datastore         Perform commands to set up the datastore.
db                Perform various tasks on the database.
front-end-build   Creates and minifies css and JavaScript files
generate          Generate empty extension files to expand CKAN
jobs              Manage background jobs
less              Compile all root less documents into their CSS counterparts
minify            Create minified versions of the given Javascript and CSS files.
notify            Send out modification notifications.
plugin-info       Provide info on installed plugins.
profile           Code speed profiler
search-index      Creates a search index for all datasets
seed              Create test data in the database.
server            Start Development server.
sysadmin          Gives sysadmin rights to a named user.
tracking          Update tracking statistics.
translation       Translation helper functions
user              Manage users.
views             Create views on relevant resources
================= ============================================================

asset: WebAssets commands
==================================

Usage::

>     asset build            - Builds bundles, regardless of whether they are changed or not

Reviewer: I think it would be helpful to put the full command-line for these commands e.g.
```suggestion
    ckan asset build            - Builds bundles, regardless of whether they are changed or not
```
For brevity I think we can get away with not specifying the `-c` parameter in this .rst file, because it will prompt you for that, and it's explained earlier in this .rst.

    search-index clear [DATASET_NAME]   - clears the search index for the provided dataset or for the whole ckan instance

seed: Create test data in the database
==================================

Examples::

      ckan -c /path/to/ckan.ini seed 
      
      
server: Start Development server
==================================

Usage::

    server --host (-h)          - Set Host
    server --port (-p)          - Set Port
    server --reloader (-r)      - Use reloader
    

sysadmin: Give sysadmin rights
==============================

Gives sysadmin rights to a named user. This means the user can perform any action on any object.

For example, to make a user called 'admin' into a sysadmin::

 ckan -c /path/to/ckan.ini --plugin=ckan sysadmin add admin 

tracking: Update tracking statistics
====================================

Usage::

    tracking update [start_date]       - update tracking stats
    tracking export FILE [start_date]  - export tracking stats to a csv file

translation: Translation helper functions
===================================

Usage::

    trans js        - generate the JavaScript translations

>     trans mangle    - mangle the zh_TW translations for testing

Reviewer: It's `translation` rather than `trans`

Reviewer: Also it misses the `check-po` sub-command which will "Check po files for common mistakes"

.. note::

    Since version 2.7 the JavaScript translation files are automatically
    regenerated if necessary when CKAN is started. Hence you usually do not
    need to run ``ckan trans js`` manually.

.. _cli-user:

user: Create and manage users
=============================

Lets you create, remove, list and manage users.

For example, to create a new user called 'admin'::

>  ckan -c /path/to/ckan.ini --plugin=ckan user add admin 

Reviewer: ```suggestion
 ckan -c |ckan.ini| user add admin 
```

Reviewer: Ah I see @smotornyuk previously mentioned all the `--plugin` parameters - maybe do a global search?

.. _cli:

======================
Command Line Interface (CLI)
======================

NB: From CKAN 2.9 onwards the CKAN configuration file is named 'ckan.ini'. Previous names: 'production.ini' and 'development.ini' (plus others) may also still appear in documentation and the software. These legacy names will eventually be phased out

Most common CKAN administration tasks can be carried out from the command line

> on the server that CKAN is installed on, using the ``ckan`` command.

Reviewer: Maybe mention it replaces the 'paster' commands from CKAN 2.9 onwards

If you have trouble running ckan commands, see
`Troubleshooting ckan Commands`_ below.

.. note::

   Once you activate your CKAN virtualenv the "ckan" command is available from within any 
   location within the host environment.

   To run a ckan command without activating the virtualenv first, you have
   to give the full path the ckan script within the virtualenv, for example:

   .. parsed-literal::

      |virtualenv|/bin/ckan -c |/path/to/ckan.ini| user list 

   In the example commands below, we assume you're running the commands with
   your virtualenv activated and from your ckan directory.

The general form of a CKAN ``ckan`` command is:

.. parsed-literal::

>    ckan --config=\ |/path/to/ckan.ini| **command** 

Reviewer: I think you need to define `|/path/to/ckan.ini|` in this file: https://github.com/ckan/ckan/blob/master/doc/conf.py#L39 (I could be wrong!) or just do `/path/to/ckan.ini`.

And I think it might be easier for users if we give the default path to the file, as before, rather than `/path/to`. On both developer and server installs everyone just uses the default path, afaict.

Author: OK thanks David - Iv'e updated the doco to say "generally the location of the config file will be at /etc/ckan/default/ directory, however it can be located in _any_ directory on the host machine"

Reviewer: That's a useful explainer. However I still think that in all other places in the docs, it should be `/etc/ckan/default/ckan.ini`, not `/path/to/ckan.ini`. So everyone can just copy and paste the command into their shell.

In addition, ckan.ini needs adding to conf.py, otherwise the pipe characters render (and they shouldn't):
<img width="321" alt="Screenshot 2020-02-07 at 11 51 18" src="https://user-images.githubusercontent.com/307612/74027705-94e74180-49a0-11ea-9ba0-10fdbe3a116e.png">

Reviewer: And the conf.py problem is causing warnings on build:
```
/vagrant/src/ckan/doc/maintaining/getting-started.rst:81: WARNING: Undefined substitution referenced: "ckan.ini".
/vagrant/src/ckan/doc/maintaining/installing/install-from-package.rst:110: WARNING: Undefined substitution referenced: "ckan.ini".
/vagrant/src/ckan/doc/maintaining/installing/install-from-package.rst:120: WARNING: Undefined substitution referenced: "ckan.ini".
/vagrant/src/ckan/doc/maintaining/installing/install-from-source.rst:259: WARNING: Undefined substitution referenced: "ckan.ini".
/vagrant/src/ckan/doc/maintaining/installing/install-from-source.rst:450: WARNING: Undefined substitution referenced: "ckan.ini".
/vagrant/src/ckan/doc/maintaining/installing/install-from-source.rst:484: WARNING: Undefined substitution referenced: "ckan.ini".
```

Author: OK in the ../doc/conf.py file I have added this line: ".. |path/to/ckan.ini| replace:: |config_dir|/ckan.ini"

The `` --config`` option tells CKAN where to find your config file, which it
reads for example to know which database it should use. As you'll see in the
examples below, this option can be given as ``-c`` for short.

``command`` should be replaced with the name of the CKAN command that you wish
to execute. Most commands have their own subcommands and options. 

.. note::

  You may also specify the location of your config file using the CKAN_INI
  environment variable. You will no longer need to use --config= or -c to
  tell ckan where the config file is:

  .. parsed-literal::

     export CKAN_INI=\ |/path/to/ckan.ini|
     

>   You can still run the ckan command in the same directory as the CKAN config file when the config file is named 'development.ini'. You will not be required to use --config or -c in this case. This will eventually be phased out in further releases of CKAN

Reviewer: ```suggestion
  You can still run the ckan command in the same directory as the CKAN config file when the config file is named 'development.ini'. You will not be required to use --config or -c in this case. However this usage is deprecated and will be phased out in a future CKAN release.
```

.. _cli:

======================
Command Line Interface (CLI)
======================

NB: From CKAN 2.9 onwards the CKAN configuration file is named 'ckan.ini'. Previous names: 'production.ini', 'development.ini' plus others may also still appear in documentation and the software. These legacy names will eventually be phased out

Most common CKAN administration tasks can be carried out from the command line
on the server that CKAN is installed on, using the ``ckan`` command.

If you have trouble running ckan commands, see
`Troubleshooting ckan Commands`_ below.

.. note::

   Before running a CKAN ``ckan`` command, you have to activate your CKAN

>    virtualenv and change to the ``ckan``  directory, for example:

Reviewer 2: Running commands from CKAN's folder is not required now. As long, as the virtual environment is active, `ckan` command is available everywhere in the host system, even at `/`.

   .. parsed-literal::

      |activate|
      cd |virtualenv|/src/ckan

   To run a ckan command without activating the virtualenv first, you have
   to give the full path the ckan script within the virtualenv, for example:

   .. parsed-literal::

      |virtualenv|/bin/ckan -c |/path/to/ckan.ini| --plugin=ckan user list 

>    To run a ckan command without changing to the ckan directory first, add

Reviewer 2: As mentioned above, `ckan` doesn't use `--plugin` and available globally with active virtual environment, so the whole paragraph can be dropped

   the ``--plugin=ckan`` option to the command. For example:

   .. parsed-literal::

      ckan -c |/path/to/ckan.ini| --plugin=ckan user list 

   In the example commands below, we assume you're running the commands with
   your virtualenv activated and from your ckan directory.

The general form of a CKAN ``ckan`` command is:

.. parsed-literal::

   ckan --config=\ |/path/to/ckan.ini| **command** 

The ``--config`` option tells CKAN where to find your config file, which it
reads for example to know which database it should use. As you'll see in the
examples below, this option can be given as ``-c`` for short.

``command`` should be replaced with the name of the CKAN command that you wish
to execute. Most commands have their own subcommands and options. For example,

> to print out a list of all of your CKAN site's users do:

Reviewer 2: it seems that here is some missing example

  environment variable. You will no longer need to use --config= or -c to
  tell ckan where the config file is:

  .. parsed-literal::

     export CKAN_INI=\ |/path/to/ckan.ini|
     
  You can still run the ckan command in the same directory as the CKAN config file when the config file is named 'development.ini'. You will not be required to use --config or -c in this case. However this usage is deprecated and will be phased out in a future CKAN release.
  
  .. parsed-literal::

     |cd /path/to/ckandir;| ckan command

.. parsed-literal::

Commands and Subcommands

   ckan -c |/path/to/ckan.ini| user list

(Here ``user`` is the name of the CKAN command you're running, and ``list`` is
a subcommand of ``user``.)

For a list of all available commands, you will need to access `CKAN Commands Reference`_. 

Each command has its own help text, which tells you what subcommands and
options it has (if any). To print out a command's help text, run the command
with the ``--help`` option::

   ckan -c |/path/to/ckan.ini| user --help

-------------------------------
Troubleshooting ckan Commands
-------------------------------

Permission Error
================

> If you receive 'Permission Denied' error, try running ckan with sudo.

Reviewer: I'm not clear that this is good advice. The postgres commands should be run as 'postgres' user. But none others that I'm aware.

Reviewer: This is existing advice. No point digging into this now.
